### DISCUSSION TITLE
PyWorker - reasons to have it and reasons to drop it

### DISCUSSION THREAD

**Discussion by WebReflection** [pyscript/pyscript#1790] [General]
As it cannot be consistently available in both *main* and *worker* worlds (or at least that's what we suggest in our code, I haven't even tried) it's clear that *PyWorker* might cause more issues than it solves:

  * reaching it out in the *main* thread means the interpreter already bootstrapped in there ... what's the use case there?
  * reaching it out in the *worker* thread throws an error because branching workers logic is convoluted and weird ... no use case there (?!) but also we're already in a worker ... there's nothing more non-blocking than that
  * reaching it out in the *main* thread via JavaScript, allowing to use any arbitrary type (currently `py` or `mpy`) has tons of use cases that have been already demoed and shown here and there ... as JS on main requires zero bootstrap but having it orchestrating or reaching out foreign PLs is definitively desirable

Accordingly, my proposal about *PyWorker* is the following:

  * we just don't expose it on the *Python* side of affair as to have workers the `worker` attribute is there for that very same reason
  * we still export *PyWorker* for the *JS* use case out of the module, so that more advanced use cases become possible, separating main/pure Web expectations with the *PyScript* workers "*wonder*" we offer

This proposal requires changes but I think all of these are pretty straight forward and welcome for easier docs around *PyWorker* without compromising the *JS* only story on the *main* page for slightly more advanced users ... is there anyone against it?

**Vote** with :+1: this discussion to signal your approval, use :-1: otherwise to signal your disapproval, hopefully commenting what is it that you don't like (or why) and propose a way forward so we can work either ways around this topic, thank you!

**Comment by fpliger** [comment]
>> we just don't expose it on the Python side of affair as to have workers the worker attribute is there for that very same reason

I'm not sure I understand this sentence. Can you elaborate?

>> we still export PyWorker for the JS use case out of the module, so that more advanced use cases become possible, separating main/pure Web expectations with the PyScript workers "wonder" we offer

So, what you are proposing here is to only supporting it in JS? 


Overall I feel like I'm lacking to understand the proposal and how the concrete new API would look like. Could you give a couple of examples of "what happens today" and "what it'd look like in the future", and "this would not be supported anymore because this happens"? (or something like this)

**Reply by WebReflection** [reply]
basically `from pyscript import PyWorker` will throw as we don't expose it there ... *but* ...

```html
<script type="module">
import { whenDefined } from '@pyscript/core';
whenDefined('py').then(({ PyWorker }) => {
  const pw = new PyWorker(...args);
});
// or ...
whenDefined('mpy').then(({ PyWorker }) => {
  const pw = new PyWorker(...args);
});
</script>
```

This is for JS only and the fact it's async is because plugins are loaded on demand ... no Python code will have a PyWorker but procedural JS can still bootstrap a PyWorker if needed and use zero `<py-script>` or `<script type="py">` on the page, still fully using the power of Python in a worker.

Does this explain any better the idea/proposal?
